plugins/samples/domain_rewrite/plugin.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START serviceextensions_plugin_domain_rewrite]
+use lol_html::*;
+use proxy_wasm::traits::*;
+use proxy_wasm::types::*;
+use std::cell::RefCell;
+use std::error::Error;
+use std::rc::Rc;
+
+proxy_wasm::main! {{
+    proxy_wasm::set_log_level(LogLevel::Trace);
+    proxy_wasm::set_http_context(|_, _| -> Box<dyn HttpContext> { MyHttpContext::new()});
+}}
+
+struct MyHttpContext<'a> {
+    // Some member variables need to be wrapped in Rc<Refcell<T>> so that we can
+    // modify the variable in multiple places as well as share ownership. By
+    // using Rc, we are commiting to not modfying the variable in multiple places
+    // simultaneously, otherwise the code will panic and plugin will crash.
+
+    // Stores HTML as rewriter parses and modifies HTML.
+    // Only stores completed sections(e.g., when rewriter parses "<di", output
+    // will be empty until the next chunk comes in. If the next chunk was
+    // "v> <h1>foo</h", the output will then contain "<div><h1>foo").
+    output: Rc<RefCell<Vec<u8>>>,
+    // HTML rewriter. Member of MyHttpContext a.k.a "StreamContext" so that the
+    // rewriter persists across multiple body callbacks.
+    rewriter: Option<HtmlRewriter<'a, Box<dyn FnMut(&[u8])>>>,
+}
+
+impl<'a> MyHttpContext<'a> {
```

Comment:
Do we need these lifetime annotations?
https://doc.rust-lang.org/rust-by-example/scope/lifetime/explicit.html

I'm out of my depth here. Maybe we should find a seasoned Rustacean to do a style review?

Reply:
When a reference is part of a struct definition, we need explicit lifetime annotations. https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/second-edition/ch10-03-lifetime-syntax.html#lifetime-annotations-in-struct-definitions

Though it isn't obvious here, HtmlRewriter relies on a reference to Event handlers and thus needs an explicit lifetime annotation. https://docs.rs/lol_html/2.2.0/lol_html/struct.HtmlRewriter.html https://docs.rs/lol_html/2.2.0/lol_html/trait.HandlerTypes.html#associatedtype.ElementHandler

Reply:
Proxying Rust reviewer feedback:

"I also wonder if you really need the lifetime variable. Your pub fn new can produce a MyHttpContext of any lifetime, which usually means it might as well always use the 'static lifetime."

"finally, it's unusual to have fn new return Box<Self>. The caller can add a box themselves if they need it."

Reply:
We need the lifetime variable because static isn't a good fit for our use. We don't want MyHttpContext to share a lifetime with the program. If we wanted the rewriter to share the lifetime of the plugin, static would be a better fit.

Reply:
Yeah, I think clarifying that lifetime matches MyHttpContext makes sense. Can you address the feedback about returning a Box?